cmd/ris-mirror/rismirror/router.go

```diff
@@ -52,17 +55,62 @@ func (r *Router) GetVRFs() []*vrf.VRF {
 }
 
 func (r *Router) addVRF(rd uint64, sources []*grpc.ClientConn) {
-	v := r.vrfRegistry.CreateVRFIfNotExists(vrf.RouteDistinguisherHumanReadable(rd), rd)
+	r.vrfsMu.Lock()
+	defer r.vrfsMu.Unlock()
+
+	// cleanup old ris clients of the vrf, before freshly initializing the vrf with new connections
+	r.stopVRF(rd)
 
-	r.vrfs[rd] = newVRF(v.IPv4UnicastRIB(), v.IPv6UnicastRIB())
+	v := r.vrfRegistry.CreateVRFIfNotExists(vrf.RouteDistinguisherHumanReadable(rd), rd)
+	r.vrfs[rd] = newVRF(v)
 
+	// init all ris clients freshly
 	for _, src := range sources {
-		r.connectVRF(rd, src, 4)
-		r.connectVRF(rd, src, 6)
+		rc4 := r.connectVRF(rd, src, 4)
+		rc6 := r.connectVRF(rd, src, 6)
+		// we cannot call v directly here, because the ris clients are only in the specific implementation of the vrf struct.
+		r.vrfs[rd].Clients = append(r.vrfs[rd].Clients, rc4, rc6)
+	}
+}
+
+func (r *Router) removeVRF(rd uint64) {
+	r.vrfsMu.Lock()
+	defer r.vrfsMu.Unlock()
+	// for removal, first stop all vrfs
+	r.stopVRF(rd)
+	// then we clean the map entries
+	vrf := r.vrfs[rd]
+	r.vrfRegistry.UnregisterVRF(vrf.vrf)
+	delete(r.vrfs, rd)
+}
+
+func (r *Router) dropAllVRFs() {
+	for _, vrf := range r.GetVRFs() {
+		r.removeVRF(vrf.RD())
+	}
+}
+
+func (r *Router) stopVRF(rd uint64) {
+	// no lock here, because this method may be only called with locked vrfMu!
+
+	// check if we even have clients for this rd
+	if _, exists := r.vrfs[rd]; !exists {
```

Comment:
I think this check is not necessary. The for loop below doesn't do anything when the entry does not exist.

Reply:
The for loop cycles over ` r.vrfs[rd].Clients` - so we get a null pointer if `r.vrfs[rd]` is nil i'd say.

Reply:
No. Iterating over a nil slice is fine. It just doesn't iterate.
https://play.golang.com/p/SKpZkFaPdAx

Reply:
Oh, now I see. Why not something like this then:
```
v := r.vrfs[rd]
if v == nil {
   return
}

for _, rc := range v.Clients {
...
```